### ISSUE TITLE
TestCopyOnWriteUpdate::testUpdateWithoutCondition is flaky

### ISSUE THREAD

**Issue by szehon-ho** [apache/iceberg#4031]
https://github.com/apache/iceberg/blob/master/spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestUpdate.java#L222

Example failure : https://github.com/apache/iceberg/runs/5030269431?check_suite_focus=true

```
org.apache.iceberg.spark.extensions.TestCopyOnWriteUpdate > testUpdateWithoutCondition[catalogName = spark_catalog, implementation = org.apache.iceberg.spark.SparkSessionCatalog, config = {type=hive, default-namespace=default, clients=1, parquet-enabled=false, cache-enabled=false}, format = avro, vectorized = false, distributionMode = range] FAILED
    org.junit.ComparisonFailure: Added data files count must match expected:<[3]> but was:<[2]>
        at org.junit.Assert.assertEquals(Assert.java:117)
        at org.apache.iceberg.spark.extensions.SparkRowLevelOperationsTestBase.validateSnapshot(SparkRowLevelOperationsTestBase.java:213)
        at org.apache.iceberg.spark.extensions.SparkRowLevelOperationsTestBase.validateCopyOnWrite(SparkRowLevelOperationsTestBase.java:193)
        at org.apache.iceberg.spark.extensions.TestUpdate.testUpdateWithoutCondition(TestUpdate.java:222)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)

```

It seems the hash function is not deterministic and sometimes two rows does hash into same file with low probability.

Not familiar if any deterministic way to test, wondering if we just assert that deletedFiles == addedDataFiles, and that this value <= 3, instead of exactly 3?  Open to suggestions

**Comment by rdblue** [comment]
What hash function is not deterministic? That seems odd to me. Are we generating random data or something?

**Comment by szehon-ho** [comment]
Yea , I meant there looks like some randomness of Spark shuffle assigning the row to same/different Spark partition..  I'll have to investigate if I have some time

**Comment by aokolnychyi** [comment]
@szehon-ho @rdblue, I think the problem is that we request a hash distribution by `_file` and sometimes records for multiple files end up in a single output task. That leads to an inconsistent number of output files. It used to happen frequently when we had only 4 shuffle partitions but then I changed it to 200. Looks like this is still not enough and we should come up with a more reliable fix.

```
    // set the num of shuffle partitions to 200 instead of default 4 to reduce the chance of hashing
    // records for multiple source files to one writing task (needed for a predictable num of output files)
    withSQLConf(ImmutableMap.of(SQLConf.SHUFFLE_PARTITIONS().key(), "200"), () -> {
      sql("UPDATE %s SET id = -1", tableName);
    });
```

Let me see the PR.

**Comment by aokolnychyi** [comment]
I commented on the PR.

**Comment by rdblue** [comment]
Got it. So the problem is that Spark's sampling isn't deterministic. That makes sense to me.

**Comment by aokolnychyi** [comment]
Resolving as it was fixed in PR #4033.
